### DISCUSSION TITLE
Access Prisma binaries without network

### DISCUSSION THREAD

**Discussion by tlap9** [prisma/prisma#17496] [Q&A] [answered]
### Bug description

Hey, i'm trying to run 'npx prisma migrate deploy' in a CI pipeline who doesn't have network but Prisma want  absolutely to go search binaries online. 

Here is the Dockerfile :

```
FROM base/debian-apt:buster as configApt
FROM base/npm-configuration:latest as npmConfig

FROM node:lts as ts-compiler 
WORKDIR /usr/app 

COPY --from=configApt / /
RUN apt-get update && apt-get install --no-install-recommends --yes openssl

COPY --from=npmConfig / /
RUN npm i -g prisma
COPY package*.json ./
COPY tsconfig.json ./
COPY prisma/ ./ 
COPY .env ./ 

RUN npm install
COPY . ./ 
RUN npm run build

FROM node:lts as ts-remover 
WORKDIR /usr/app
COPY --from=ts-compiler /usr/app/package*.json ./ 
COPY --from=ts-compiler /usr/app/dist ./
COPY --from=ts-compiler /usr/app/prisma/ ./ 
COPY --from=ts-compiler /usr/app/.env ./
RUN npm install --omit=dev

FROM node:lts 
WORKDIR /usr/app
COPY --from=ts-remover /usr/app ./ 
RUN npx prisma migrate deploy
EXPOSE 8000
CMD npm run start:prod
```

### Expected behavior

Prisma to go search local binaries when he can't access network.

### Prisma information

```prisma
// Add your schema.prisma
```
generator client {
  provider = "prisma-client-js"
  binaryTargets = ["prisma-fmt-debian-openssl-3.0.x"]
}

### Environment & setup

- OS: WSL / Debian
- Database: PostgreSQL
- Node.js version: 16.18
- CI: Concourse

### Prisma Version

```
4.9.0
```

**Comment by nurul3101** [comment]
Hello @tom-laplace 👋 

Have you tried adding environment variables for [Custom Engine File locations](https://www.prisma.io/docs/reference/api-reference/environment-variables-reference#custom-engine-file-locations)?
Can you try adding them, you can specify the custom path of local binaries through these variables.

**Reply by tlap9** [reply]
Ok I tried things, I think i'm kinda close now :

```
COPY package*.json ./
COPY tsconfig.json ./
COPY prisma/ ./ 
COPY .env ./ 
RUN npm install 
COPY . ./  

ENV PRISMA_CLI_QUERY_ENGINE_TYPE binary
ENV PRISMA_INTROSPECTION_ENGINE_BINARY=/usr/app/prisma_engines/debian-introspection-engine
ENV PRISMA_MIGRATION_ENGINE_BINARY=/usr/app/prisma_engines/debian-migration-engine
ENV PRISMA_QUERY_ENGINE_BINARY=/usr/app/prisma_engines/debian-query-engine

RUN chmod +x /usr/app/prisma_engines/debian-introspection-engine
RUN chmod +x /usr/app/prisma_engines/debian-migration-engine
RUN chmod +x /usr/app/prisma_engines/debian-query-engine

RUN npx prisma -v
RUN npx prisma generate
RUN npm run build
```

But i have this error at npx prisma -v

```
Error: Command failed with exit code 2: /usr/app/prisma_engines/debian-query-engine --version
/usr/app/prisma_engines/debian-query-engine: 1: ���XS��: not found
/usr/app/prisma_engines/debian-query-engine: 2: Syntax error: word unexpected (expecting ")")
    at makeError (/usr/app/node_modules/prisma/build/index.js:3641:18)
    at handlePromise (/usr/app/node_modules/prisma/build/index.js:4409:33)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async getEngineVersion (/usr/app/node_modules/prisma/build/index.js:89905:20) {
  shortMessage: 'Command failed with exit code 2: /usr/app/prisma_engines/debian-query-engine --version',
  command: '/usr/app/prisma_engines/debian-query-engine --version',
  escapedCommand: '"/usr/app/prisma_engines/debian-query-engine" --version',
  exitCode: 2,
  signal: undefined,
  signalDescription: undefined,
  stdout: '',
  stderr: '/usr/app/prisma_engines/debian-query-engine: 1: \x1F�\b���\x07XS��\x07: not found\n' +
    '/usr/app/prisma_engines/debian-query-engine: 2: Syntax error: word unexpected (expecting ")")',
  failed: true,
  timedOut: false,
  isCanceled: false,
  killed: false
}  
```

It's the same for the 2 others.

To have local binaries I did :

curl -L https://binaries.prisma.sh/2.0.x/7d77a7cd570c9dedb4cdf31a1cbec615cf2e1eda/debian-openssl-1.0.x/query-engine.gz --output debian-query-engine

Do we have an other way to get the binaries I need ?

**Reply by nurul3101** [reply]
Using `binaries.prisma.sh` is the official way to get the binaries.
I was able to find a related Stack Overflow Answer for the error that you are getting, can you have a look? https://stackoverflow.com/questions/36725864/syntax-error-word-unexpected-expecting

**Reply by tlap9** [reply]
Ok it's finally working !! 

I'm going to document how I did it tomorrow in case others people need this use case.

I learned a lot about how Prisma work in the past few days, thanks a lot

**Reply by nurul3101** [reply]
I am glad to hear that! Documenting would definitely help a lot of people! 🙂

**Comment by cuttingedge1109** [comment]
`prisma generate` normally requires query engine and schema engine. You have to download the binaries and place in a certain place(ofc, need to unzip the gz file).
```sh
export PRISMA_CLI_QUERY_ENGINE_TYPE=binary
export PRISMA_SCHEMA_ENGINE_BINARY=${schema_engine_binary_path}
export PRISMA_QUERY_ENGINE_LIBRARY=${query_engine_binary_path}
export PRISMA_QUERY_ENGINE_BINARY=${query_engine_binary_path}
```

Also if prisma requires introspection and migration binaries for its job, need to set these.
```sh
export PRISMA_INTROSPECTION_ENGINE_BINARY=${introspection_engine_binary_path}
export PRISMA_MIGRATION_ENGINE_BINARY=${migration_engine_binary_path}
```

**Comment by ludralph** [comment, marked as answer]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
